fixit/cli/add_new_rule.py

```diff
@@ -113,5 +113,37 @@ def main() -> None:
     create_rule_file(file_path, rule_name)
 
 
+def register_subparser(parsers: argparse._SubParsersAction) -> None:
+    """Add subparser for `add_new_rule` command."""
+    add_new_rule_parser = parsers.add_parser(
+        "add_new_rule",
+        help="Creates a skeleton of adding new rule file.",
+    )
+
+    add_new_rule_parser.add_argument(
+        "-p",
+        "--path",
+        type=is_path_exists,
+        default=Path("fixit/rules/new.py"),
+        help="Path to add rule file, defaults to fixit/rules/new.py",
+    )
+
+    add_new_rule_parser.add_argument(
+        "-n",
+        "--name",
+        type=str,
+        default="",
+        help="Name of the rule, defaults to `New`",
+    )
+
+    add_new_rule_parser.set_defaults(subparser_fn=_main)
+
+
+def _main(args: argparse.Namespace) -> None:
+    file_path = args.path
+    rule_name = args.name if args.name else file_path.stem
+    create_rule_file(file_path, rule_name)
+
+
 if __name__ == "__main__":
     main()
```

Comment:
We can simply move the parser initialization here and then use `register_subparser` to initialize it.
That way we can remove the existing `main()` which is pretty much duplicated now.

```
parser = argparse.ArgumentParser(
        description="Creates a skeleton of adding new rule file"
    )
```


The same approach also applies to other scripts.

Reply:
As per current implementation, `register_subparser(parsers: argparse._SubParsersAction)` takes `subparser` object and add arguments to it and as per your suggestion, `register_subparser(parser)` will take `ArgumentParser ()` object. 

Can you please help me with how you want to achieve this in order to support backward compatibility? and to remove existing `main()` which contains duplicate arguments.

Reply:
Both `argparse._SubParsersAction` and `ArgumentParser` have `add_parser` method.
So you can define the method as
```
from typing import Union
def register_subparser(parsers: Union[argparse._SubParsersAction, argparse.ArgumentParser]) -> None:
```

Reply:
`ArgumentParser` has no `add_parser()` method.
This is the one way I can achieve and we can still remove the existing `main()` which is duplicate. But, this way there is duplicate arguments (`add_argument()`)

Let me know if you have any suggestions.

```py
def register_subparser(parser: Union[argparse._SubParsersAction, argparse.ArgumentParser]):

    if isinstance(parser, argparse._SubParsersAction):
        add_new_rule_parser = parser.add_parser(
            "add_new_rule",
            help="Creates a skeleton of adding new rule file.",
        )
        add_new_rule_parser.add_argument( "--path", ..)
        add_new_rule_parser.add_argument( "--name", ..)

        add_new_rule_parser.set_defaults(subparser_fn=_main)

    else:
        parser.add_argument( "--path", ..)
        parser.add_argument( "--name", ..)

        args = parser.parse_args()
        _main(args)
```

```py
def _main(args: argparse.Namespace) -> None:
    file_path = args.path
    rule_name = args.name if args.name else file_path.stem
    create_rule_file(file_path, rule_name)
```

Reply:
Sorry, I read it wrong. This is my suggestion:
So the common parts are those `add_argument` statement and the `set_defaults`.
If we simply let `register_subparser` take an optional `_SubParsersAction`, and it calls `argparse.ArgumentParser` to create a parser if parsers is `None`. Then we can remove existing duplicated `main()`.
Each `if __name__ == "__main__":` can simply calls `parser = register_subparser(); args = parser.parse_args(argv); args.subparser_fn(args)`

Reply:
How about this one? 

```py
def add_rule_parser():
    parser = argparse.ArgumentParser(
        description="Creates a skeleton of adding new rule file",
    )

    parser.add_argument( "--path" ..)
    parser.add_argument( "--name" ..)

    return parser


def register_subparser(parser: argparse._SubParsersAction=None) -> None:
    """Add subparser for `add_new_rule` command."""
    if parser is None:
        rule_parser = add_rule_parser()
        _main(rule_parser.parse_args())

    else:
        sub_parser = parser.add_parser(
            "add_new_rule",
            description="Creates a skeleton of adding new rule file.",
            help="Creates a skeleton of adding new rule file.",
        )
        sub_parser.add_argument( "--path" ..)
        sub_parser.add_argument( "--name" ..)
        
        sub_parser.set_defaults(subparser_fn=_main)


def _main(args: argparse.Namespace) -> None:
    file_path = args.path
    rule_name = args.name if args.name else file_path.stem
    create_rule_file(file_path, rule_name)


if __name__ == "__main__":
    register_subparser()
```

Reply:
The `add_argument` parts are still duplicated which can be refactored as a function to be reused.